lib/cucumber/tcl/framework.tcl

```diff
@@ -12,31 +12,42 @@ set STEPS [list]
 #
 # Define procs to match Gherkin keyworkds that put data in the STEPS array
 #
-proc Given {re body} {
-  global STEPS
-  lappend STEPS [list $re $body]
+proc Given args {
+  _add_step {*}$args
 }
 
-proc When {re body} {
-  global STEPS
-  lappend STEPS [list $re $body]
+proc When args {
+  _add_step {*}$args
 }
 
-proc Then {re body} {
-  global STEPS
-  lappend STEPS [list $re $body]
+proc Then args {
+  _add_step {*}$args
 }
 
-proc And {re body} {
+proc And args {
+  _add_step {*}$args
+}
+
+proc _add_step args {
+
   global STEPS
-  lappend STEPS [list $re $body]
+
+  if {[llength $args] == 2} {
+    set re [lindex $args 0]
+    set params {}
+    set body [lindex $args 1]
+  } elseif {[llength $args] == 3} {
+    set re [lindex $args 0]
+    set params [lindex $args 1]
+    set body [lindex $args 2]
+  }
```

Comment:
I don't pretend to understand this code completely, but I get the impression it would only work for max 3 parameters, is that right? Can you see a way to make it more generic?